katdal/concatdata.py

```diff
@@ -200,44 +201,78 @@ def _initial_dtype(self):
 
 
+def infer_dtypes(sensor_data_sequence):
+    """List all known dtypes found in a sequence of sensor data objects.
+
+    This extracts the dtypes from a sequence of sensor data objects and
+    finds the unique set (where string types of various lengths are represented
+    by the string type with the maximum length). If all the objects have
+    unknown dtype, return [None] instead.
+
+    XXX The list of dtypes can be further collapsed to a single dtype if needed
+    via NumPy coercion. This will effectively forgo any relative type checking.
+
+    Parameters
+    ----------
+    sensor_data_sequence : sequence of sensor data objects
+        These objects may include :class:`SensorData`, :class:`numpy.ndarray`,
+        :class:`CategoricalData` and lists of sensor values
+
+    Returns
+    -------
+    dtypes : list of :class:`numpy.dtype` objects, or [None]
+        List of unique dtypes found in sequence, or [None] if dtype is unknown
+
+    """
+    dtypes = [infer_dtype(sd) for sd in sensor_data_sequence]
+    # Ignore unavailable dtypes unless nothing is available
+    dtypes = [dt for dt in dtypes if dt is not None]
+    dtypes = unique_in_order(dtypes) if dtypes else [None]
+    if all(np.issubdtype(dtype, np.str) for dtype in dtypes):
+        # Strings of different lengths have different dtypes
+        # (e.g. '|S1' vs '|S10') but can be safely concatenated
+        dtypes = [np.dtype('|S%d' % (max([dt.itemsize for dt in dtypes]),))]
+    return dtypes
+
+
 class ConcatenatedSensorData(SensorData):
     """The concatenation of multiple raw (uncached) sensor data sets.
 
     This is a convenient container for returning raw (uncached) sensor data sets
     from a :class:`ConcatenatedSensorCache` object. It only accesses the
     underlying data sets when explicitly asked to via the __getitem__ interface,
-    but provides quick access to metadata such as sensor dtype, name and number
-    of data points.
+    but provides quick access to metadata such as sensor dtype and name.
 
     Parameters
     ----------
-    data : sequence of recarray-like with fields ('timestamp', 'value', 'status')
-        Uncached sensor data as a list of record arrays or equivalent (such as a
-        :class:`h5py.Dataset`)
+    data : sequence of recarray-like with fields 'timestamp', 'value' and optionally 'status'
+        Uncached sensor data as a list of record arrays or equivalent (such as
+        an :class:`h5py.Dataset`)
 
     """
+
     def __init__(self, data):
-        self.data = data
+        self._data = data
         names = unique_in_order([sd.name for sd in data])
         if len(names) != 1:
-            raise ConcatenationError('Cannot concatenate sensors with different names: %s' % (names,))
+            raise ConcatenationError('Cannot concatenate sensor %r with different '
+                                     'underlying names: %s' % (self.name, names))
```

Comment:
True. I was confused by the high-level sensor name (the key in the SensorCache) and the low-level / underlying sensor name (SensorData.name). I wanted to refer to the former, but that is not available at this level. Fixed.